Add SCTP Network Policy tests

fedepaol opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test

/kind feature

> /kind flake

**What this PR does / why we need it**:

This PR completes the checklist from [https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/0015-20180614-SCTP-support.md#sctp-connectivity-tests](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/0015-20180614-SCTP-support.md#sctp-connectivity-tests), testing network policies with SCTP.

This is achieved by adding SCTP support to both the `agnhost` `connect` command and `porter` command, and to use then in the network policy tests to check if network policies are working.
This is complementary to https://github.com/kubernetes/kubernetes/pull/87926

In order to maintain the ` [Feature:SCTPConnectivity]` and have the sctp tests optional, I started converting the current policy tests in a tabular form. To do that (and to maintain sctp optional), I removed the `BeforeEach` function and converted it to a custom one invoked with the different protocols in each table body.
Not sure if this is the right approach but I could not find a better one that prevents duplication while still allowing us to skip the SCTP tests.


**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

NONE

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->

[KEP] : https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/0015-20180614-SCTP-support.md

```

fedepaol commented:
/cc @danwinship 
/cc @vishnoianil

I still need to complete the conversion to tables, but since it's pretty long I stopped waiting for feedbacks.

fedepaol commented:
/cc @janosi

fedepaol commented:
/hold

danwinship commented:
/assign @danwinship

cmluciano reviewed (commented):
/ok-to-test

aojea commented:
> In order to maintain the ` [Feature:SCTPConnectivity]` and have the sctp tests optional, I started converting the current policy tests in a tabular form. To do that (and to maintain sctp optional), I removed the `BeforeEach` function and converted it to a custom one invoked with the different protocols in each table body.
> Not sure if this is the right approach but I could not find a better one that prevents duplication while still allowing us to skip the SCTP tests.

just my 2 cents, I had very bad experiences with Ginkgo trying to do things "out" of its way. I'd would try to aggregate the common code in functions and duplicate all the tests for each protocol 🤷‍♂

fedepaol commented:
> > In order to maintain the ` [Feature:SCTPConnectivity]` and have the sctp tests optional, I started converting the current policy tests in a tabular form. To do that (and to maintain sctp optional), I removed the `BeforeEach` function and converted it to a custom one invoked with the different protocols in each table body.
> > Not sure if this is the right approach but I could not find a better one that prevents duplication while still allowing us to skip the SCTP tests.
> 
> just my 2 cents, I had very bad experiences with Ginkgo trying to do things "out" of its way. I'd would try to aggregate the common code in functions and duplicate all the tests for each protocol man_shrugging

Thanks for your suggestion, it was one of the possible alternatives I had in mind while writing this.
I'm fine with either of those, but I'll wait for other opinions so there is consensus before starting to refactor. The infrastructure is all there so it's "just" a matter of how I do rearrange the code.

fedepaol commented:
@aojea I tried to rearrange under your suggestion, refactoring only the first test to avoid to do unnecessary work. Does this approach sound more reasonable?
I am afraid the pr won't be as easy to review as it was before code is getting moved around.

Others? @danwinship maybe?

fedepaol committed:
Add tests for SCTP network policies.

Signed-off-by: Federico Paolinelli <fpaoline@redhat.com>

danwinship commented on test/images/agnhost/connect/connect.go:
```diff
@@ -81,3 +95,26 @@ func main(cmd *cobra.Command, args []string) {
+		os.Exit(1)
+	}
+
+	conn, err := sctp.DialSCTPExt("sctp", nil, addr, sctp.InitMsg{MaxInitTimeout: uint16(timeout.Milliseconds())})
```
This doesn't seem to be correct; `MaxInitTimeout` → `sinit_max_init_timeo` is not a timeout on the connect operation as a whole, it's an upper bound on what the timeout can be for each attempt, of which there are 8 by default, starting at 3s each and then doubling until they hit the max. (eg, if you pass `--timeout 15` you'll get 3s + 6s + 12s + 15s + 15s + 15s + 15s + 15s = 1m36s).

There doesn't seem to be any good way to make the operation itself time out when you want it to without breaking the protocol so I think you're going to need to rewrite this to just do a timeout-less connect in a goroutine, with a separate timer.

fedepaol replied:
Thanks for pointing out, I completely mis-interpreted it. I'll go with the goroutine approach.

danwinship commented on test/images/agnhost/porter/porter.go:
```diff
@@ -98,3 +106,38 @@ func serveTLSPort(port, value string) {
+		if err != nil {
+			log.Fatal("Failed to accept SCTP. error:", err)
+		}
+		go func(conn *sctp.SCTPConn) {
```
defer conn.Close()

danwinship reviewed (commented):
reviewing just the agnhost parts, which I think you should split out to a separate PR now

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -55,50 +76,83 @@ var _ = SIGDescribe("NetworkPolicy [LinuxOnly]", func() {
+		if protocol == tcp {
+			podServer, service = createServerPodAndService(f, f.Namespace, fmt.Sprintf("server-%s", protocol), []int{80, 81})
+		} else {
+			podServer, service = createServerPodAndService(f, f.Namespace, fmt.Sprintf("server-%s", protocol), []int{}, []sctpPort{sctpPort{80, 82}, sctpPort{81, 83}}...)
```
so as discussed on slack yesterday, I'd avoid major rewrites to the NetworkPolicy code. Just add one or two new SCTP tests and don't change the existing tests.

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -1688,39 +1851,68 @@ func cleanupServerPodAndService(f *framework.Framework, pod *v1.Pod, service *v1
-				"pod-name": podName,
+func createNetworkClientPodWithRestartPolicy(f *framework.Framework, namespace *v1.Namespace, podName string, targetService *v1.Service, targetPort int, protocol connProtocol, restartPolicy v1.RestartPolicy) *v1.Pod {
+	switch protocol {
+	case tcp:
```
it seems like you ought to be able to share a lot more code between the tcp and sctp cases here

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -1688,39 +1851,68 @@ func cleanupServerPodAndService(f *framework.Framework, pod *v1.Pod, service *v1
+							"/bin/sh",
+							"-c",
+							fmt.Sprintf("for i in $(seq 1 5); do nc -vz -w 8 %s.%s %d && exit 0 || sleep 1; done; exit 1",
+								targetService.Name, targetService.Namespace, targetPort),
```
just use `agnhost connect` for both the tcp and sctp cases

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -1688,39 +1851,68 @@ func cleanupServerPodAndService(f *framework.Framework, pod *v1.Pod, service *v1
+						Command: []string{"/bin/sh"},
+						Args: []string{
+							"-c",
+							fmt.Sprintf("for i in $(seq 1 5); do /agnhost connect %s.%s:%d --protocol sctp --timeout 1s && exit 0 || sleep 1; done; exit 1", targetService.Name, targetService.Namespace, targetPort),
```
`--timeout 1s` here is inconsistent with the TCP case which uses 8 seconds

fedepaol commented:
Removed the agnhost section as it's now being done in https://github.com/kubernetes/kubernetes/pull/90797

Also, this PR will need the agnhost image to be promoted before it can land.

fedepaol commented:
/hold cancel

Tested on a local OCP cluster with ovn-kubernetes as sdn:
[https://pastebin.com/jiqkvW4N](https://pastebin.com/jiqkvW4N)

aojea commented:
/retest

aojea commented:
it LGTM
is the agnost image ` "2.20"` already available @fedepaol ?

fedepaol commented:
> it LGTM
> is the agnost image ` "2.20"` already available @fedepaol ?

Actually 2.21. I tested with it. And thanks to your comment I realized I forgot to force push the last change :P
Promoted in https://github.com/kubernetes/k8s.io/pull/983

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -1890,10 +2081,15 @@ func collectPodsAndNetworkPolicies(f *framework.Framework, podClient *v1.Pod) ([
 // Will also assign a pod label with key: "pod-name" and label set to the given podName for later use by the network
 // policy.
-func createServerPodAndService(f *framework.Framework, namespace *v1.Namespace, podName string, ports []int) (*v1.Pod, *v1.Service) {
+func createServerPodAndService(f *framework.Framework, namespace *v1.Namespace, podName string, ports []int, sctpPorts ...sctpPort) (*v1.Pod, *v1.Service) {
```
change `ports` to `tcpPorts`

or have a single `ports` argument that takes a struct of port and protocol

fedepaol replied:
Changed. I'd prefer to keep the variadic function so `sctpPort`s are optional and do not impact existing tests.

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -1785,9 +1804,180 @@ var _ = SIGDescribe("NetworkPolicy [LinuxOnly]", func() {
+	ginkgo.Context("NetworkPolicy between server and client", func() {
+		ginkgo.BeforeEach(func() {
+			ginkgo.By("Creating a simple server that serves on port 80 and 81.")
+			podServer, service = createServerPodAndService(f, f.Namespace, "server", []int{80, 81}, []sctpPort{{80, 82}, {81, 83}}...)
```
This creates a server that's listening on both TCP and SCTP but you never use the TCP ports

fedepaol commented on test/e2e/network/network_policy.go:
```diff
@@ -46,6 +46,26 @@ connections from one of the clients. The test then asserts that the clients
+	sctp connProtocol = "sctp"
+)
+
+func (p connProtocol) v1Protocol() *v1.Protocol {
```
Yes, I can't remember the original version (it was some time ago) but probably it's a leftover of a more complex approach to make all the existing tests viable for both protocols.

fedepaol commented on test/e2e/network/network_policy.go:
```diff
@@ -1937,6 +2133,50 @@ func createServerPodAndService(f *framework.Framework, namespace *v1.Namespace,
+					HTTPGet: &v1.HTTPGetAction{
+						Path: "/",
+						Port: intstr.IntOrString{
+							IntVal: int32(sctpPort.readinessPort),
```
Brilliant! Thanks for the suggestion

fedepaol commented:
@danwinship changed, PTAL

fedepaol commented:
Validation here https://pastebin.com/ta1cyKkv

fedepaol commented on test/e2e/network/network_policy.go:
```diff
@@ -46,6 +46,16 @@ connections from one of the clients. The test then asserts that the clients
+
+const (
+	tcp  connProtocol = "tcp"
+	sctp connProtocol = "sctp"
```
I have it to have a type safe parameter to pass where I need to choose between tcp and sctp.
In `func testCanConnectProtocol(f *framework.Framework, ns *v1.Namespace, podName string, service *v1.Service, targetPort int, protocol connProtocol) {` for example. I can replace with a bare string if you prefer.

fedepaol commented on test/e2e/network/network_policy.go:
```diff
@@ -1785,9 +1794,180 @@ var _ = SIGDescribe("NetworkPolicy [LinuxOnly]", func() {
+	ginkgo.Context("NetworkPolicy between server and client using SCTP", func() {
+		ginkgo.BeforeEach(func() {
+			ginkgo.By("Creating a simple server that serves on port 80 and 81.")
+			podServer, service = createServerPodAndService(f, f.Namespace, "server", []int{}, []int{80, 81}...)
```
Np, agree it's cleaner that way. I'll change it.

fedepaol commented on test/e2e/network/network_policy.go:
```diff
@@ -1937,6 +2118,42 @@ func createServerPodAndService(f *framework.Framework, namespace *v1.Namespace,
+			ReadinessProbe: &v1.Probe{
+				Handler: v1.Handler{
+					Exec: &v1.ExecAction{
+						Command: []string{"/agnhost", "connect", "--protocol=sctp", "--timeout 1s", fmt.Sprintf("127.0.0.1:%d", port)},
```
It won't be working without the timeout as the sctp implementation uses a select.

danwinship replied:
oh... actually, that's a bug in the agnhost code; if you don't specify a timeout, it treats that as "time out immediately", but it should be "time out never" like in the TCP case

fedepaol replied:
ok, let me fix it then

fedepaol replied:
opened https://github.com/kubernetes/kubernetes/pull/92657

fedepaol commented:
Ok, I think things are converging. @danwinship please take another look.
I managed to put some more code together by using the same readiness probe for both protocols.

fedepaol commented:
/hold
Depends on https://github.com/kubernetes/kubernetes/pull/92657

danwinship commented:
> /hold
> Depends on #92657

That might take a while. I'm fine merging this now if you add back the theoretically-unnecessary `--timeout` flag

fedepaol commented:
> > /hold
> > Depends on #92657
> 
> That might take a while. I'm fine merging this now if you add back the theoretically-unnecessary `--timeout` flag

Done! Sorry I missed this last comment, it was buried in CI reports and I was monitoring the other PR.

fedepaol commented:
/hold cancel

fedepaol commented:
/retest

fedepaol commented:
For getting approved the image bump up:
/cc @luxas @mkumatag @listx 
as they are listed as approvers in `test/utils/image/OWNERS` 

/cc @BenTheElder as he bumped the previous agnhost version.

fedepaol commented:
> For getting approved the image bump up:
> /cc @luxas @mkumatag @listx
> as they are listed as approvers in `test/utils/image/OWNERS`
> 
> /cc @BenTheElder as he bumped the previous agnhost version.

@luxas @mkumatag @listx ping!

spiffxp commented:
/test pull-kubernetes-e2e-kind-ipv6

spiffxp commented on test/e2e/network/network_policy.go:
```diff
@@ -2004,20 +2202,18 @@ func createNetworkClientPodWithRestartPolicy(f *framework.Framework, namespace *
 						"-c",
-						fmt.Sprintf("for i in $(seq 1 5); do nc -vz -w 8 %s.%s %d && exit 0 || sleep 1; done; exit 1",
-							targetService.Name, targetService.Namespace, targetPort),
+						fmt.Sprintf("for i in $(seq 1 5); do /agnhost connect %s.%s:%d --protocol %s --timeout 8s && exit 0 || sleep 1; done; exit 1", targetService.Name, targetService.Namespace, targetPort, connectProtocol),
```
Comments that would help future test maintainers: What's the rationale behind the magic numbers here?
- why retry 5 times and not more/less
- why timeout at 8s instead of more/less

fedepaol replied:
Those magic numbers were already there, I left them while changing the way we open the socket in a way that supports both TCP and SCTP.

I tried to look back up to https://github.com/kubernetes/kubernetes/commit/3471daf7599604561825c7133aee6cc76408d199 but could not find a proper reasoning behind that.

spiffxp replied:
Fair enough, thanks for looking

spiffxp commented on test/e2e/network/network_policy.go:
```diff
@@ -1785,9 +1791,180 @@ var _ = SIGDescribe("NetworkPolicy [LinuxOnly]", func() {
 	})
 })
 
+var _ = SIGDescribe("NetworkPolicy [Feature:SCTPConnectivity][LinuxOnly]", func() {
```
According to https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/0015-20180614-SCTP-support.md#sctp-connectivity-tests these should be marked as `[Disruptive]`?

danwinship replied:
Yes. And we should belatedly fix the other SCTPConnectivity tests to have that now too

fedepaol replied:
Done, and added a separate commit to mark the old ones disruptive as well.

fedepaol committed:
Mark SCTPConnectivity tests as disruptive.

Since the SCTP module verification tests were added, their result may be affected by
running the SCTPConnectivity tests. For this reason, they are now marked as disruptive.

Signed-off-by: Federico Paolinelli <fpaoline@redhat.com>

fedepaol commented:
/retest

spiffxp reviewed (approved):
/approve
/lgtm

spiffxp commented:
/approve cancel
actually... wait

https://github.com/kubernetes/kubernetes/pull/94152 landed which bumped agnhost to 2.22, I am concerned this will move it back to 2.21

spiffxp commented on build/dependencies.yaml:
```diff
@@ -10,7 +10,7 @@ dependencies:
   # then after merge and successful postsubmit image push / promotion, bump this
   - name: "agnhost: dependents"
-    version: "2.20"
+    version: "2.21"
```
I believe if you rebase you'll find this is now 2.22

spiffxp replied:
🤦 you did rebase, and I missed the context above, nevermind, this is fine

spiffxp commented on test/utils/image/manifest.go:
```diff
@@ -200,7 +200,7 @@ const (
 func initImageConfigs() map[int]Config {
 	configs := map[int]Config{}
-	configs[Agnhost] = Config{promoterE2eRegistry, "agnhost", "2.20"}
+	configs[Agnhost] = Config{promoterE2eRegistry, "agnhost", "2.21"}
```
```suggestion
	configs[Agnhost] = Config{promoterE2eRegistry, "agnhost", "2.22"}
```

spiffxp replied:
nevermind, this isn't strictly speaking necessary

spiffxp commented:
/approve
ok, don't mind me, this is fine

fedepaol commented:
> /approve
> ok, don't mind me, this is fine

:slightly_smiling_face:

spiffxp commented:
/milestone v1.20
This is a test-only change; the tests introduced are gated behind Disruptive.

The main that concerns me from a will-this-be-disruptive perspective is the bump to agnhost:2.21.  But we're going to have to bump it eventually, so may as well do it now.  Also it looks like that bump would only pickup https://github.com/kubernetes/kubernetes/pull/90797 which looks unobjectionable to me
